### PR TITLE
[Workspace] Dashboard admin(groups/users) implementation

### PR DESCRIPTION
**Pull request by yubonluo**
### Description
This PR set the backend roles in groups or users, whoever has the backend roles or exactly match the user or groups ids defined in this yml.config/dynamic config will be regard as dashboard admin.
Dashboard admin will have the access to all the workspaces and objects inside OpenSearch Dashboards.

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6389

## Changelog
 - feat: [Workspace] Dashboard admin(groups/users) implementation.
## Screenshot

No UI change in this PR.

### Yml config test
1. Clone the latest code and clone security-dashboards-plugin to the plugins directory
2. Build the security-dashboards-plugin and bootstrap
3. Download the patch file [test.patch](https://github.com/opensearch-project/OpenSearch-Dashboards/files/14975232/test.patch) and yarn `git apply ./test.patch`
4. Modify `config/opensearch_dashboards.yml`
```yml
workspace.enabled: true 
opensearch.hosts: ["https://localhost:9200"]
opensearch.username: "username"
opensearch.password: "password" 
opensearch.ssl.verificationMode: none
savedObjects.permission.enabled: true

opensearchDashboards.dashboardAdmin.groups: ["dashboard_admin"]
opensearchDashboards.dashboardAdmin.users: ["dashboard_admin"]
```
5. Run `yarn start --no-base-path` and visit http://localhost:5601
7. You can check whether it is dashboards admin in the console.
image

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/workspace/server/utils.ts [hunk 1]
```diff
@@ -50,3 +54,63 @@ export const getPrincipalsFromRequest = (
...
+  let adminConfigArray;
+  try {
+    adminConfigArray = JSON.parse(adminConfig);
+  } catch (e) {
+    return [];
+  }
+  return adminConfigArray;
+};
+
+export const getApplicationOSDAdminConfig = async (
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/utils.ts:91]:
```suggestion
export const getOSDAdminConfigFromApplicationConfig = async (
```
What about we changing the function name to like this?

File: src/plugins/workspace/server/utils.ts [hunk 2]
```diff
@@ -50,3 +54,63 @@ export const getPrincipalsFromRequest = (
...
+      .catch(() => undefined),
+    applicationConfigClient
+      .getEntityConfig('opensearchDashboards.dashboardAdmin.users')
+      .catch(() => undefined),
+  ]);
+
+  return [stringToArray(groupsResult), stringToArray(usersResult)];
+};
+
+export const getOSDAdminConfig = async (globalConfig$: Observable<SharedGlobalConfig>) => {
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/utils.ts:109]:
```suggestion
export const getOSDAdminConfigFromYMLConfig = async (globalConfig$: Observable<SharedGlobalConfig>) => {
```

File: src/plugins/workspace/server/utils.ts [hunk 3]
```diff
@@ -50,3 +54,63 @@ export const getPrincipalsFromRequest = (
...
+  const globalConfig = await globalConfig$.pipe(first()).toPromise();
+  const groupsResult = (globalConfig.opensearchDashboards?.dashboardAdmin?.groups ||
+    []) as string[];
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/utils.ts:112]:
```suggestion
    []);
```
Is the type assertion `as string[]` necessary here?

File: src/plugins/workspace/server/utils.ts [hunk 4]
```diff
@@ -50,3 +54,63 @@ export const getPrincipalsFromRequest = (
...
+  configGroups: string[],
+  configUsers: string[]
+) => {
+  if (configGroups.length === 0 && configUsers.length === 0) {
+    updateWorkspaceState(request, {
+      isDashboardAdmin: false,
+    });
+    return;
+  }
+  const groupMatchAny = groups.some((group) => configGroups.includes(group)) || false;
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/utils.ts:71]:
```suggestion
  const groupMatchAny = groups.some((group) => configGroups.includes(group));
```
groups.some will return boolean value, what is the `|| false` used for?

File: src/plugins/workspace/server/types.ts [hunk 5]
```diff
@@ -128,6 +129,9 @@ export interface AuthInfo {
   user_name?: string;
 }
 
+export interface AppPluginSetupDependencies {
+  applicationConfig: ApplicationConfigPluginSetup;
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/types.ts:133]:
```suggestion
  applicationConfig?: ApplicationConfigPluginSetup;
```
applicationConfig is an optional dependency of workspace.

File: src/plugins/workspace/server/utils.ts (hunk 1 above)
Comment (yubonluo) [reply src/plugins/workspace/server/utils.ts:91]: Sure, I will rename this function.

File: src/plugins/workspace/server/utils.ts (hunk 2 above)
Comment (yubonluo) [reply src/plugins/workspace/server/utils.ts:109]: Sure, I will rename this function.

File: src/plugins/workspace/server/utils.ts (hunk 3 above)
Comment (yubonluo) [reply src/plugins/workspace/server/utils.ts:112]: Yes, `as string[]` can make the return type consistent when calling the method in `src/plugins/workspace/server/plugin.ts`(106 line).

File: src/plugins/workspace/server/utils.ts (hunk 4 above)
Comment (yubonluo) [reply src/plugins/workspace/server/utils.ts:71]: I will delete the redundant code.

File: src/plugins/workspace/server/plugin.ts
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/plugin.ts:1]: Seems some lines are not covered by unit test, can we do that?

**Review by SuZhou-Joe — APPROVED** [review]

File: src/plugins/workspace/server/plugin.ts
Comment (yubonluo) [reply src/plugins/workspace/server/plugin.ts:1]: Sure, I will add some unit tests to cover these lines.

**Review by SuZhou-Joe — APPROVED** [review]

File: src/plugins/workspace/server/utils.ts [hunk 6]
```diff
@@ -50,3 +54,34 @@ export const getPrincipalsFromRequest = (
 
   throw new Error('UNEXPECTED_AUTHORIZATION_STATUS');
 };
+
+export const updateDashboardAdminStateForRequest = (
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/utils.ts:58]: @ruanyl and I think perhaps we should set `isDashboardAdmin` to true if groups.length === 0 && users.length === 0. They will be empty when permissionControl flag is open but authentication is not enabled, in that case, everyone should be dashboard admin.

Comment (yubonluo) [reply src/plugins/workspace/server/utils.ts:58]: Sure, the logic has been updated. If security plugin uninstall and permission control enabled, login defaults to OSD admin.

**Review by ruanyl — APPROVED** [review]

**Review by SuZhou-Joe — APPROVED** [review]

**Pull request merged**